diff --git a/include/modules/clock.hpp b/include/modules/clock.hpp
index b3e2634..e22a3a6 100644
--- a/include/modules/clock.hpp
+++ b/include/modules/clock.hpp
@@ -22,6 +22,7 @@ class Clock : public ALabel {
   Clock(const std::string&, const Json::Value&);
   ~Clock() = default;
   auto update() -> void;
+  auto handleToggle(GdkEventButton *const &e) -> bool;
 
  private:
   util::SleeperThread thread_;
@@ -30,13 +31,25 @@ class Clock : public ALabel {
   bool fixed_time_zone_;
   int time_zone_idx_;
   date::year_month_day cached_calendar_ymd_ = date::January/1/0;
-  std::string cached_calendar_text_;
+  struct Calendar {
+    std::string header;
+    std::string text;
+  };
+  Calendar cached_calendar_;
+  static constexpr int months_in_year_ = 12;
+  std::array<const char *, months_in_year_> month_names_;
+  bool tooltip_has_calendar_;
+  int calendar_month_offset_;
 
   bool handleScroll(GdkEventScroll* e);
 
-  auto calendar_text(const waybar_time& wtime) -> std::string;
-  auto weekdays_header(const date::weekday& first_dow, std::ostream& os) -> void;
-  auto first_day_of_week() -> date::weekday;
+  template<typename M>
+  auto calendar(const waybar_time& wtime, M&& month_name) -> Calendar;
+  template<typename M>
+  auto header(const date::year_month& ym, M&& month_name) -> std::string const;
+  auto calendar_text(const date::year_month_day& ymd) -> std::string const;
+  auto weekdays_header(const date::weekday& first_dow, std::ostream& os) -> void const;
+  auto first_day_of_week() -> date::weekday const;
 };
 
 }  // namespace waybar::modules
diff --git a/resources/config b/resources/config
index 13dc94c..371eec7 100644
--- a/resources/config
+++ b/resources/config
@@ -65,7 +65,8 @@
     },
     "clock": {
         // "timezone": "America/New_York",
-        "tooltip-format": "<big>{:%Y %B}</big>\n<tt><small>{calendar}</small></tt>",
+        "format": "{:%a %m/%d %I:%M %p}",
+        "tooltip-format": "<big>{calendar_header}</big>\n<tt><small>{calendar}</small></tt>",
         "format-alt": "{:%Y-%m-%d}"
     },
     "cpu": {
diff --git a/src/modules/clock.cpp b/src/modules/clock.cpp
index 22bedc7..aae7612 100644
--- a/src/modules/clock.cpp
+++ b/src/modules/clock.cpp
@@ -11,10 +11,29 @@
 #include <locale.h>
 #endif
 
+namespace {
+
+bool tooltip_has_calendar(const Json::Value& config) {
+  if (!config["tooltip-format"].isString()) return false;
+  const auto tooltip_format = config["tooltip-format"].asString();
+  bool contains_calendar = (tooltip_format.find("{calendar}") != std::string::npos) ||
+    (tooltip_format.find("{calendar_header}") != std::string::npos);
+  return contains_calendar;
+}
+
+}
+
 using waybar::modules::waybar_time;
 
+template <auto fn>
+using deleter_from_fn = std::integral_constant<decltype(fn), fn>;
+
+template <typename T, auto fn>
+using deleting_unique_ptr = std::unique_ptr<T, deleter_from_fn<fn>>;
+
 waybar::modules::Clock::Clock(const std::string& id, const Json::Value& config)
-    : ALabel(config, "clock", id, "{:%H:%M}", 60, false, false, true), fixed_time_zone_(false) {
+    : ALabel(config, "clock", id, "{:%H:%M}", 60, false, false, true), fixed_time_zone_(false),
+      tooltip_has_calendar_(tooltipEnabled() && tooltip_has_calendar(config)), calendar_month_offset_(0) {
   if (config_["timezone"].isString()) {
     spdlog::warn("As using a timezone, some format args may be missing as the date library havn't got a release since 2018.");
     time_zone_ = date::locate_zone(config_["timezone"].asString());
@@ -27,6 +46,21 @@ waybar::modules::Clock::Clock(const std::string& id, const Json::Value& config)
     locale_ = std::locale("");
   }
 
+  {
+    // Load full month names from glibc.  Date format %B does not work for month name.
+    deleting_unique_ptr<std::remove_pointer<locale_t>::type, freelocale>
+      posix_locale{newlocale(LC_ALL, locale_.name().c_str(), nullptr)};
+    if (posix_locale) {
+      for (int i = 0; i < months_in_year_; ++i) {
+        month_names_[i] = nl_langinfo_l(ALTMON_1 + i, posix_locale.get());
+      }
+    } else {
+      for (int i = 0; i < months_in_year_; ++i) {
+        month_names_[i] = nl_langinfo(ALTMON_1 + i);
+      }
+    }
+  }
+
   thread_ = [this] {
     dp.emit();
     auto now = std::chrono::system_clock::now();
@@ -60,9 +94,12 @@ auto waybar::modules::Clock::update() -> void {
 
   if (tooltipEnabled()) {
     if (config_["tooltip-format"].isString()) {
-      const auto calendar = calendar_text(wtime);
       auto       tooltip_format = config_["tooltip-format"].asString();
-      auto       tooltip_text = fmt::format(tooltip_format, wtime, fmt::arg("calendar", calendar));
+      const auto cal = calendar(wtime, [this](int month_idx) {
+        return std::string(month_names_[month_idx]);
+      });
+      auto       tooltip_text = fmt::format(tooltip_format, wtime,
+        fmt::arg("calendar_header", cal.header), fmt::arg("calendar", cal.text));
       label_.set_tooltip_markup(tooltip_text);
     } else {
       label_.set_tooltip_markup(text);
@@ -104,14 +141,57 @@ bool waybar::modules::Clock::handleScroll(GdkEventScroll *e) {
   return true;
 }
 
-auto waybar::modules::Clock::calendar_text(const waybar_time& wtime) -> std::string {
+auto waybar::modules::Clock::handleToggle(GdkEventButton* const& e) -> bool {
+  if (tooltip_has_calendar_) {
+    if (e->type == GDK_BUTTON_PRESS) {
+      switch (e->button) {
+        case 1: --calendar_month_offset_; break;
+        case 2: calendar_month_offset_ = 0; break;
+        case 3: ++calendar_month_offset_; break;
+      }
+      update();
+    }
+    return true;
+  } else {
+    return ALabel::handleToggle(e);
+  }
+}
+
+template<typename M>
+auto waybar::modules::Clock::calendar(const waybar_time& wtime, M&& month_name) -> Calendar {
   const auto daypoint = date::floor<date::days>(wtime.ztime.get_local_time());
   const auto ymd = date::year_month_day(daypoint);
+
+  if (calendar_month_offset_ != 0) {
+    const date::year_month curr_month = ymd.year()/ymd.month();
+    const date::year_month offset_month = (calendar_month_offset_ > 0) ?
+      (curr_month + date::months(calendar_month_offset_)) :
+      (curr_month - date::months(-calendar_month_offset_));
+    auto head = header(offset_month, month_name);
+    auto body_text = calendar_text(offset_month/0);
+    return Calendar{head, body_text};
+  }
+
   if (cached_calendar_ymd_ == ymd) {
-    return cached_calendar_text_;
+    return cached_calendar_;
   }
 
-  const date::year_month ym(ymd.year(), ymd.month());
+  auto head = header(ymd.year()/ymd.month(), month_name);
+  const Calendar calendar{head, calendar_text(ymd)};
+  cached_calendar_ymd_ = ymd;
+  cached_calendar_ = calendar;
+  return calendar;
+}
+
+template<typename M>
+auto waybar::modules::Clock::header(const date::year_month& ym, M&& month_name) -> std::string const {
+  auto year_string = date::format(locale_, "%Y", ym.year());
+  return fmt::format("{} {}", year_string, month_name(unsigned(ym.month()) - 1));
+}
+
+// ymd.day() == day(0) means don't mark the current day.
+auto waybar::modules::Clock::calendar_text(const date::year_month_day& ymd) -> std::string const {
+  const date::year_month ym = ymd.year()/ymd.month();
   const auto             curr_day = ymd.day();
 
   std::stringstream os;
@@ -145,12 +225,12 @@ auto waybar::modules::Clock::calendar_text(const waybar_time& wtime) -> std::str
 
   auto result = os.str();
   cached_calendar_ymd_ = ymd;
-  cached_calendar_text_ = result;
+  cached_calendar_.text = result;
   return result;
 }
 
 auto waybar::modules::Clock::weekdays_header(const date::weekday& first_dow, std::ostream& os)
-    -> void {
+    -> void const {
   auto wd = first_dow;
   do {
     if (wd != first_dow) os << ' ';
@@ -168,16 +248,8 @@ auto waybar::modules::Clock::weekdays_header(const date::weekday& first_dow, std
   os << "\n";
 }
 
-#ifdef HAVE_LANGINFO_1STDAY
-template <auto fn>
-using deleter_from_fn = std::integral_constant<decltype(fn), fn>;
-
-template <typename T, auto fn>
-using deleting_unique_ptr = std::unique_ptr<T, deleter_from_fn<fn>>;
-#endif
-
 // Computations done similarly to Linux cal utility.
-auto waybar::modules::Clock::first_day_of_week() -> date::weekday {
+auto waybar::modules::Clock::first_day_of_week() -> date::weekday const {
 #ifdef HAVE_LANGINFO_1STDAY
   deleting_unique_ptr<std::remove_pointer<locale_t>::type, freelocale> posix_locale{
       newlocale(LC_ALL, locale_.name().c_str(), nullptr)};
